ipcache: Introduce the ability to inherit CIDR prefixes

gandro opened the pull request:
This PR introduces a new feature in the IPCache metadata subsystem:
The ability for prefixes with local identities to inherit CIDR labels
from encompassing prefixes. This is needed to implement
https://github.com/cilium/design-cfps/pull/17 where we want to introduce a new way to manage
identities for IPs allowlisted by DNS lookups. Without going into too
much detail, the key point of that CFP is that IPs returned by DNS
lookups will be associated with a `fqdn:` label.

Those IPs returned by a DNS lookup however might also be selected by a
CIDR prefix, and thus might also optionally need a `cidr:` label if (but
only if) a selecting CIDR policy is present.

To give an example: If we perform a DNS lookup on `one.one.one.one` on a
pod with DNS redirects enabled, we may observe IP `1.1.1.1` and thus
associated that IP with a `fqdn:` label. If at the same time, we also
have a separate CIDR policy allowing traffic to `1.0.0.0/8`, then that
IP also needs the `cidr:1.0.0.0/8` label, to make it is selected by both
the FQDN and the CIDR policy.

To implement this, the IPCache metadata subsystem is slightly augmented:

1. Whenever metadata for a prefix is upserted or removed, we no longer
   enqueue only the prefix it self, but also all potentially affected
   prefixes or IPs contained by it. This allows `InjectLabels` to then
   down-propagate the changes to any child prefixes.
2. In `resolveIdentity` (called by `InjectLabels`), we check if there
   are any parent prefixes with a CIDR label that we want to inherit.

At the moment, only local identities without a `reserved:` label may
inherit a CIDR label from a parent prefix. This currently therefore only
applies to identities containing `fqdn:` labels. Going forward, this
infrastructure might however also be useful to implement non-K8s label
sources.

gandro committed:
ipcache: Remove unused GetMetadataLabelsByIP function

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Remove direct access to PrefixInfo

This commit removes the direct access to `PrefixInfo` as it is not
needed for the current caller in the node manager (which only requires
the source). A subsequent commit will introduce additional labels to a
prefix not found in the `PrefixInfo`. Therefore, to avoid callers
observing incomplete information, we hide the direct access.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Unexport PrefixInfo

There are no longer is access outside of the IPCache package.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
labels: Add HasSource helper

This helper allows to check if a label set contains a label from a
particular source.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
identity: Add FQDN label source

This label source will superseed CIDR labels for identities generated by
ToFQDN rules. Like CIDR labels, FQDN labels are node local identities.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Use canonical prefix in metadata map

A IP prefix may have multiple representations. For example
`192.168.0.0/24` can also be represented as `192.168.50.60/24` (a IPv4
prefix using a non-masked address) or `::ffff:192.168.0.0/24` (a
IPv4-mapped IPv6 address which will use the IPv4 stack on the host).
Those prefixes might look different, but they all represent exactly the
same destination. To avoid issues where callers might provide metadata
for a prefix in non-canonical form, this PR introduces a helper which
translates every IP used for upserts, deletions and lookups into the
canonical form before the metadata map is accessed.

This commit does not fix any known bug and merely introduces this in the
spirit of defensive programming.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Fix outdated source code comment

The comment was written is no longer valid, as CIDR policies now also
funnel their prefixes through the IPCache metadata subsystem.

However, the logic needs to remain, as other callers (e.g. the node
manager) now can inject multiple prefixes that share the same identity.
The code comment is updated to reflect that.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Introduce the ability to inherit CIDR prefixes

This commit introduces a new feature in the IPCache metadata subsystem:
The ability for prefixes with local identities to inherit CIDR labels
from encompassing prefixes. This is needed to implement
cilium/design-cfps#17 where we want to introduce a new way to manage
identities for IPs allowlisted by DNS lookups. Without going into too
much detail, the key point of that CFP is that IPs returned by DNS
lookups will be associated with a `fqdn:` label.

Those IPs returned by a DNS lookup however might also be selected by a
CIDR prefix, and thus might also optionally need a `cidr:` label if (but
only if) a selecting CIDR policy is present.

To give an example: If we perform a DNS lookup on `one.one.one.one` on a
pod with DNS redirects enabled, we may observe IP `1.1.1.1` and thus
associated that IP with a `fqdn:` label. If at the same time, we also
have a separate CIDR policy allowing traffic to `1.0.0.0/8`, then that
IP also needs the `cidr:1.0.0.0/8` label, to make it is selected by both
the FQDN and the CIDR policy.

To implement this, the IPCache metadata subsystem is slightly augmented:

1. Whenever metadata for a prefix is upserted or removed, we no longer
   enqueue only the prefix it self, but also all potentially affected
   prefixes or IPs contained by it. This allows `InjectLabels` to then
   down-propagate the changes to any child prefixes.
2. In `resolveIdentity` (called by `InjectLabels`), we check if there
   are any parent prefixes with a CIDR label that we want to inherit.

At the moment, only local identities without a `reserved:` label may
inherit a CIDR label from a parent prefix. This currently therefore only
applies to identities containing `fqdn:` labels. Going forward, this
infrastructure might however also be useful to implement non-K8s label
sources.

A note on the implementation: To find affected child prefixes, we
currently perform a linear scan over the metadata map. While this is a
`O(n)` operation, finding CIDR parents in a trie could be done in `O(log
n)`, but at the moment we expect `n` to be in hundreds to the low
thousands at most. For most real world cases, we therefore expect the
current implementation to perform just as well if not better than a
trie. Also note that the linear scan is skipped if the prefix is a
single IP and thus cannot contain any other prefixes besides itself.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro commented on pkg/ipcache/metadata.go:
```diff
@@ -565,7 +628,9 @@ func (ipc *IPCache) resolveIdentity(ctx context.Context, prefix netip.Prefix, in
+		!lbls.Has(labels.LabelIngress[labels.IDNameIngress]) &&
+		!lbls.HasSource(labels.LabelSourceFQDN) &&
+		!lbls.HasSource(labels.LabelSourceCIDR) {
 		cidrLabels := labels.GetCIDRLabels(prefix)
```
note: This case _should_ not be hit anymore, even before this PR. But we still have unit tests relying on it. To avoid adding even more commits to this PR, I decided to not emit a "should not be reached" warning log message here and leave this for a follow-up PR.

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -214,6 +216,43 @@ func (m *metadata) getLocked(prefix netip.Prefix) prefixInfo {
+		return true
+	}
+
+	return parent.Contains(child.Addr()) && child.Bits() >= parent.Bits()
```
Why is the latter part of this expression needed (the bits check)? Isn't `Contains()` enough?

squeed replied:
Child could have the same base address but a shorter prefix. So you have to check the length too.

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -429,7 +433,7 @@ func (ipc *IPCache) InjectLabels(ctx context.Context, modifiedPrefixes []netip.P
 	// Recalculate policy first before upserting into the ipcache.
 	if len(idsToAdd) > 0 {
-		ipc.UpdatePolicyMaps(ctx, idsToAdd, idsToDelete)
+		ipc.UpdatePolicyMaps(ctx, idsToAdd, nil)
```
How is this related to "ipcache: Only update policy maps for new identities " commit?

squeed replied:
It’s not directly, just a small error I noticed that causes an unnecessary policy update.

christarazi replied:
Let's pull this out to a separate PR so we can review it separately

gandro replied:
Sounds good, I'll pull out the `ipcache: Only update policy maps for new identities` commit into a separate PR. 

This particular change highlighted here should be a no-op, as `idsToDelete` is empty at this call site.

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -262,14 +307,15 @@ func (ipc *IPCache) InjectLabels(ctx context.Context, modifiedPrefixes []netip.P
 		} else {
 			// Insert to propagate the updated set of labels after removal.
-			newID, _, err = ipc.resolveIdentity(ctx, prefix, prefixInfo, prefixInfo.RequestedIdentity().ID())
+			newID, isNew, err = ipc.resolveIdentity(ctx, prefix, prefixInfo, prefixInfo.RequestedIdentity().ID())
```
Is this optimization related to this PR as in it was enabled by the previous commits in this PR? Or is this something that we can just cherry pick as a separate thing?

squeed replied:
At first I was going to say yes, but actually you’re right; it is always possible that label injection does not allocate a new identity, and thus this case applies independently.

christarazi replied:
Same as https://github.com/cilium/cilium/pull/32578#discussion_r1605345758

nathanjsweet reviewed (approved):
Very cool!

christarazi reviewed (commented):
Could you elaborate on 

> A note on the implementation: To find affected child prefixes, we
currently perform a linear scan over the metadata map. While this is a
`O(n)` operation, finding CIDR parents in a trie could be done in `O(log
n)`, but at the moment we expect `n` to be in hundreds to the low
thousands at most. For most real world cases, we therefore expect the
current implementation to perform just as well if not better than a
trie. Also note that the linear scan is skipped if the prefix is a
single IP and thus cannot contain any other prefixes besides itself.

What is part of `n` where we expect it to be in the hundreds or low thousands? What if we add pod IPs into this as https://github.com/cilium/cilium/pull/31409 would do?

gandro committed:
ipcache/test: Add benchmark comparing hashmap and trie

This adds a benchmark testing the overhead of using a naive linear scan
over a Go map vs using a bitlpm trie.

The benchmark harness generates a number of child (leaf) prefixes, plus
a number of parent (non-leaf) prefixes.  To simulate somewhat realistic
behavior, we don't allocate IPs randomly across the whole IP space, but
rather from a contiguous block of IPs with a given sparseness factor
(i.e. a factor of 1.0 will allocate every single IP out of the block, a
factor of 0.1 will allocate only 10% of the available IP space).

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

christarazi reviewed (approved):
Thanks for adding the benchmark! I agree with not blocking the PR. I was trying to understand just how much the impact was in case we faced issues with users and how much work it would be to move towards the trie as the mitigation.
